crates/runtime/src/flight.rs

```diff
@@ -283,6 +274,34 @@ fn handle_datafusion_error(e: DataFusionError) -> Status {
     }
 }
 
+#[allow(clippy::needless_pass_by_value)]
+fn handle_sql_datafusion_error(e: DataFusionError) -> Status {
+    match e {
+        DataFusionError::External(ref external_error) => {
+            if let Some(execution_error) =
+                external_error.to_string().strip_prefix("Execution error: ")
+            {
+                return Status::invalid_argument(execution_error.to_string());
+            }
+            Status::invalid_argument(e.to_string())
+        }
+        DataFusionError::ArrowError(ref arrow_error, _) => {
+            // The S3 query error is managed here to avoid
+            // introducing custom downcast logic within Datafusion or Object Store.
+            let error_message = format!("{arrow_error:?}");
```

Comment:
Can you put an example of what this is supposed to be matching on?

Reply:
The error pattern matched is: ArrowError(ExternalError(ObjectStore(Generic { store: "S3", source: reqwest::Error { kind: Decode, source: reqwest::Error { kind: Body, source: TimedOut } } })), None)